### PR TITLE
Plot Selector Widget - initial functionality

### LINKED ISSUE #22477: Add plot selector widget to the workbench
The idea of this is to add a widget with a list of plots, allowing an easy way to switch between them.

See https://github.com/mantidproject/documents/pull/56 for the full design document.

This first issue will add the following functions:
- List of plots shown (names are the figure titles)
- Plot list can be filtered via text box
- Plots can be closed via close button or delete key

### PR DESCRIPTION
**Pull request by ianbush**
**Description of work**

This is an initial implementation of the work described in this design document - https://github.com/mantidproject/documents/blob/master/Design/Workbench/Plot-Selector/design-plot-selector.md. This PR is open to review the code structure and check the basic functionality. Other features will be added in further PRs.

Current functionality:
- List of plots shown (names are the figure titles)
- Plot list can be filtered via text box
- Plots can be closed via close button or delete key 

**To test:**

Build and start the workbench. Make lots of plots, try closing and opening them and check nothing breaks. Does filtering work as expected?

Code review - check the MVP implementation here is sensible.

Fixes #22477 .

**Release Notes** 
*Workbench release notes will be gathered together separately*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/python/mantidqt/widgets/plotselector/widget.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        layout.sizeHint()
+        self.setLayout(layout)
+
+        # Connect presenter methods to things in the view
+        self.list_view.doubleClicked.connect(self.presenter.list_double_clicked)
+        self.filter_box.textChanged.connect(self.presenter.filter_text_changed)
+        self.close_button.clicked.connect(self.presenter.close_action_called)
+        self.keyPressed.connect(self.presenter.close_action_called)
+
+    keyPressed = Signal(int)
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/plotselector/widget.py]: Can we keep all of the class attribute definitions at the top of the class? It's bit harder to see what there is at a glance if they are scattered through the class definition.

File: qt/python/mantidqt/widgets/plotselector/widget.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    """
+    The view to the plot selector, a PyQt widget.
+    """
+    def __init__(self, presenter, parent=None):
+        """
+        Initialise a new instance of PlotSelectorWidget
+        :param presenter: The presenter that created this class
+        :param parent: A parent QWidget
+        """
+        QWidget.__init__(self, parent)
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/plotselector/widget.py]: We have used `super` elsewhere in the class, maybe use it here too for consistency?

**Review by martyngigg — COMMENTED** [review]
Code-wise this looks great - just a few minor inline comments although I would also be happy in its current state.

One general comment, that I have not been as good at either, is could we  make sure that all methods have doc strings? It would be a good habit to take forward when other people look at the code as an example.

I'll just do some functional testing before approving.

File: qt/python/mantidqt/widgets/plotselector/presenter.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+from .model import PlotSelectorModel
+from .widget import PlotSelectorWidget
+
+
+class PlotSelectorPresenter(object):
+    """
+    Presents (controls) a plot selector view. This UI element allows the user
+    to select and make active a plot.
+    """
+    def __init__(self, current_figure_class, widget = None, model = None):
```
[the remark this answers is not in the document]
Comment (ianbush) [reply qt/python/mantidqt/widgets/plotselector/presenter.py]: Yes, I think that is a good idea. I had been toying with both names, and I agree `view` makes things clear in MVP.

**Comment by ianbush** [comment]
I'll have a go at tabifying this with the workspace widget, I think that is a good idea to avoid clutter. I guess in the long run we will need to tweak the layout based on user feedback. I wonder if actually we should tabify this with Algorithms, as a user is less likely to want to use the Algoirthms widget and Plots widget at the same time, but if they are plotting may want to see the Workspace widget and Plot widget?

I agree with you on the button scaling - are you ok if I leave this to the next PR? This will have other buttons added and I can tweak the layout then.

I had forgot you can do ctrl+A, so it is not obvious! I think a Close All button is a good idea, ok if I put this in the next PR too?

**Comment by martyngigg** [comment]
Good point about wanting both docks visible - let's try it tabbed with the algorithms, I'd keep it second in the list, and see what happens.

No problem about leaving the extra visual tweaks and buttons to the next PR. :+1:

**Review by martyngigg — APPROVED** [review]
I think is great as a first implementation. Code looks good and it works as expected. :+1

**Comment by martyngigg** [comment]
Note that #22498 must be merged first.

**Pull request merged**